src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala

```diff
@@ -173,32 +192,50 @@ object CodeGenYul extends CodeGenerator {
                 new_name, // TODO rename transaction name (by adding prefix/suffix) iev: this seems to be done already
                 constructor.args.map(v => TypedName(v.varName, mapObsTypeToABI(v.typIn.toString))),
                 Seq(), //todo/iev: why is this always empty?
-                Block(constructor.body.flatMap((s: Statement) => translateStatement(s))))) //todo iev temp vars: likely a hack to work around something wrong
+                Block(constructor.body.flatMap((s: Statement) => translateStatement(s, None))))) //todo iev flatmap may be a bug to hide something wrong; None means that constructors don't return. is that true?
     }
 
     def translateTransaction(transaction: Transaction): Seq[YulStatement] = {
-        val ret: Seq[TypedName] =
-            if (transaction.retType.isEmpty) {
-                Seq()
-            } else { // TODO hard code return variable name now, need a special naming convention to avoid collisions
-                Seq(TypedName("retValTempName", mapObsTypeToABI(transaction.retType.get.toString)))
+        var id: Option[String] = None
+        val ret: Seq[TypedName] = {
+            transaction.retType match {
+                case Some(t) =>
+                    id = Some(nextRet())
+                    Seq(TypedName(id.get, mapObsTypeToABI(t.toString)))
+                case None => Seq()
             }
+        }
+
+        // update the mapping for returns
+        fnNamesToReturnArity = fnNamesToReturnArity.+((transaction.name, ret.length))
```

Comment:
Need to remove this update.

Reply:
as above: got it; i fixed this in https://github.com/mcoblenz/Obsidian/pull/331/commits/e0570da498bbec0b2c6c93ca4849e38f15321ed4 and https://github.com/mcoblenz/Obsidian/pull/331/commits/84e302ff13d644d92502141dd4df0af015e7a9f1. in retrospect this is all a bit overwrought, but it works.